src/debug/telemetry/telemetry.c

```diff
@@ -0,0 +1,143 @@
+
+#include <adsp_debug_window.h>
+#include <errno.h>
+#include <limits.h>
+#include <mem_window.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* Systic variables, one set per core */
+static int telemetry_systick_counter[CONFIG_MAX_CORE_COUNT];
+static int telemetry_prev_ccount[CONFIG_MAX_CORE_COUNT];
+static int telemetry_perf_period_sum[CONFIG_MAX_CORE_COUNT];
+static int telemetry_perf_period_cnt[CONFIG_MAX_CORE_COUNT];
+static struct telemetry_perf_queue telemetry_perf_queue[CONFIG_MAX_CORE_COUNT] = {0};
+
+static void telemetry_perf_queue_append(struct telemetry_perf_queue *q, size_t element)
+{
+	if (!q->full) {
+		q->elements[q->index] = element;
+		q->sum += element;
+		q->index++;
+		q->size++;
+		if (q->index >= AVG_PERF_MEAS_DEPTH) {
+			q->index = 0;
+			q->size = AVG_PERF_MEAS_DEPTH;
+			q->full = true;
+		}
+	} else {
+		/* no space, pop tail */
+		q->sum -= q->elements[q->index];
+		/* replace tail */
+		q->elements[q->index] = element;
+		q->sum += element;
+		/* move tail */
+		q->index++;
+		if (q->index >= AVG_PERF_MEAS_DEPTH)
+			q->index = 0;
+	}
+}
+
+static size_t telemetry_perf_queue_avg(struct telemetry_perf_queue *q)
+{
+	if (!q->size)
+		return 0;
+	return q->sum / q->size;
+}
+
+int telemetry_init(void)
+{
+	/* systick_init */
+	uint8_t slot_num = DW_TELEMETRY_SLOT;
+	volatile struct adsp_debug_window *window = ADSP_DW;
+	struct telemetry_wnd_data *wnd_data = (struct telemetry_wnd_data *)ADSP_DW->slots[slot_num];
+	struct system_tick_info *systick_info =
+			(struct system_tick_info *)wnd_data->system_tick_info;
+
+	window->descs[slot_num].type = ADSP_DW_SLOT_TELEMETRY;
+	window->descs[slot_num].resource_id = 0;
+	wnd_data->separator_1 = 0x0000C0DE;
+
+	/* Zero values per core */
+	for (int i = 0; i < CONFIG_MAX_CORE_COUNT; i++) {
+		systick_info[i].count = 0;
+		systick_info[i].last_sys_tick_count = 0;
+		systick_info[i].max_sys_tick_count = 0;
+		systick_info[i].last_ccount = 0;
+		systick_info[i].avg_utilization = 0;
+		systick_info[i].peak_utilization = 0;
+		systick_info[i].peak_utilization_4k = 0;
+		systick_info[i].peak_utilization_8k = 0;
+	}
+	return 0;
+}
+
+void telemetry_update(uint32_t begin_ccount, uint32_t current_ccount)
+{
+	int prid = cpu_get_id();
+
+	++telemetry_systick_counter[prid];
+
+	struct telemetry_wnd_data *wnd_data =
+		(struct telemetry_wnd_data *)ADSP_DW->slots[DW_TELEMETRY_SLOT];
+	struct system_tick_info *systick_info =
+		(struct system_tick_info *)wnd_data->system_tick_info;
+
+	systick_info[prid].count = telemetry_systick_counter[prid];
+	systick_info[prid].last_sys_tick_count = current_ccount - begin_ccount;
+	systick_info[prid].max_sys_tick_count =
+			MAX(current_ccount - begin_ccount,
+			    systick_info[prid].max_sys_tick_count);
+	systick_info[prid].last_ccount = current_ccount;
+
+#ifdef PERFORMANCE_MEASUREMENTS
+	const size_t measured_systick = begin_ccount - telemetry_prev_ccount[prid];
+
+	telemetry_prev_ccount[prid] = begin_ccount;
```

Comment:
This is interesting how we share the implementation with existing routines. We have similar (but not same) calculation in LL scheduler to calculate the LL load. This is done with a very lean implementation to allow it to be left on in product builds. How is this code for PERFORMANCE_MEASUREMENTS ? Is this designed to be fast enough to be left on in release builds?

Reply:
Yes, in reference firmware it was always on, so we can assume it was fast enough.